pkg/chaosdaemon/tasks/chaosonprocess.go

```diff
@@ -0,0 +1,252 @@
+// limitations under the License.
+//
+
+package tasks
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/ChaosErr"
+)
+
+var ErrCanNotAdd = errors.New("can not add")
+var ErrCanNotAssign = errors.New("can not assign")
+
+// NewChaosOnProcess init ChaosOnProcess with values can not be Assigned.
+type NewChaosOnProcess interface {
+	New(immutableValues interface{}) (ChaosOnProcess, error)
+}
+
+type AssignChaosOnProcess interface {
+	Assign(ChaosOnProcess) error
+}
+
+type TaskToProcess interface {
+	Addable
+	NewChaosOnProcess
+	AssignChaosOnProcess
+}
+
+type ChaosCanRecover interface {
+	Recover(pid PID) error
+}
+
+type ChaosOnProcess interface {
+	Inject(pid PID) error
+}
+
+// ChaosOnProcessManager is a Manager for chaos tasks base on their
+// target that marked with a PID and implement with ChaosOnProcess.
+// Since we need to :
+//	Sum tasks on same PID to one task.
+// 	Create working chaos on process called ChaosOnProcess by task info.
+// 	Assign the summed task info to the ChaosOnProcess.
+// If developers wants to use functions in ChaosOnProcessManager ,
+// their imported Task need to implement interface TaskToProcess.
+// If developers wants to totally recover task successfully
+// when no task applied on the ChaosOnProcess.
+// This ChaosOnProcess must implement ChaosCanRecover.
+// If not implement ,
+// the Recover function will return a ErrNotImplement("ChaosCanRecover") error.
+type ChaosOnProcessManager struct {
+	taskManager TaskManager
+	processMap  map[PID]ChaosOnProcess
+
+	logger logr.Logger
+}
+
+func NewChaosOnProcessManager(logger logr.Logger) ChaosOnProcessManager {
+	return ChaosOnProcessManager{
+		NewTaskManager(),
+		make(map[PID]ChaosOnProcess),
+		logger,
+	}
+}
+
+func (cm ChaosOnProcessManager) CopyTaskManager() TaskManager {
+	tm := NewTaskManager()
+	for uid, task := range cm.taskManager.TaskMap {
+		tm.TaskMap[uid] = task
+	}
+	return tm
+}
+
+func (cm ChaosOnProcessManager) CopyProcessMap() map[PID]ChaosOnProcess {
+	pm := make(map[PID]ChaosOnProcess)
+	for pid, chaosOnProcess := range cm.processMap {
+		cm.processMap[pid] = chaosOnProcess
+	}
+	return pm
+}
+
+func (cm ChaosOnProcessManager) GetWithUID(id UID) (interface{}, error) {
+	return cm.taskManager.GetWithUID(id)
+}
+
+func (cm ChaosOnProcessManager) GetWithPID(pid PID) (ChaosOnProcess, error) {
+	p, ok := cm.processMap[pid]
+	if !ok {
+		return nil, ChaosErr.NotFound("PID")
+	}
+	return p, nil
+}
+
+func (cm ChaosOnProcessManager) GetUIDsWithPID(pid PID) []UID {
+	return cm.taskManager.GetWithPID(pid)
+}
+
+// Update the task config with a same UID and PID .
+// If it comes a UID not injected , Update will return ChaosErr.NotFound("UID").
+// If it comes the import PID of task do not equal to the last one,
+// Update will return ErrUpdateTaskWithPIDChanges.
+func (cm ChaosOnProcessManager) Update(uid UID, pid PID, config TaskToProcess) error {
```

Comment:
I think it's better to add locks to operate create, update and delete.

Reply:
As a package , I think user should think about it by themselves. In bpm we add lock on process, but it finally brings dead lock in HTTP-Chaos because we do not assume bpm will do that.